chisel/src/options.rs

```
@@ -0,0 +1,127 @@
+        match self.0.get(key) {
+            Some(s) => Some(s.as_str()),
+            None => None,
+        }
+    }
+
+    /// Compares the value of `key` to the passed `val`.
+    pub fn value_eq(&self, key: &str, val: &str) -> bool {
+        match self.value_of(key) {
+            Some(s) => s == val,
+            None => false,
+        }
+    }
+
+    /// Apply all flags passed from CLI
+    pub fn apply(&mut self, matches: &ArgMatches) {
+        if matches.is_present("NO_RECOVER") {
+            self.set("util.norecover", "true");
+        }
+        if matches.is_present("VERBOSE") {
+            self.set("util.debugging", "true");
+        }
+        if let Some(value) = matches.value_of("CONFIG") {
+            self.set("run.config.path", value);
+        }
+        if let Some(values) = matches.values_of("MODULES") {
+            let values_collected = values.fold(String::new(), |mut acc, val| {
+                acc.push_str(&format!("{},", val));
+                acc
+            });
+            self.set("oneliner.modules", &values_collected);
+        }
+        if let Some(values) = matches.values_of("MODULE_OPTIONS") {
+            let values_collected = values.fold(String::new(), |mut acc, val| {
+                acc.push_str(&format!("{},", val));
+                acc
+            });
+            self.set("oneliner.modules.options", &values_collected);
+        }
+        if let Some(value) = matches.value_of("FILE") {
+            self.set("oneliner.file", value);
+        }
+        if let Some(value) = matches.value_of("OUTPUT_PATH") {
+            self.set("oneliner.output", value);
+        }
+        if let Some(value) = matches.value_of("OUTPUT_MODE") {
+            match value {
+                val @ "bin" | val @ "wat" | val @ "hex" => {
+                    self.set("output.mode", val);
+                }
+                _ => panic!("CLI parser only accepts 'bin', 'wat', or 'hex'"),
```

Why not return an error here?

---

This function does not return a set of flags, it applies arguments from `clap` to the flag set.

---

Terrible. Is there a way to signal to clap what values are accepted?

---

It does: https://docs.rs/clap/2.33.0/clap/struct.Arg.html#method.possible_values

---

Yes, in the app declaration. It should be done there, which is why it makes sense to panic here. This simply should be unreachable.

---

This panic is fine as long as the above `possible_values` will be used so that clap doesn't even let this pass through to the panic.

---

exactly